akd/src/directory.rs

```diff
@@ -409,189 +409,101 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage, H: Hasher> Directory<S, V, H> {
-                let label_for_ver = self.vrf.get_node_label::<H>(uname, false, ver).await?;
-                let non_existence_of_ver = current_azks
-                    .get_non_membership_proof(&self.storage, label_for_ver)
-                    .await?;
-                non_existence_of_future_markers.push(non_existence_of_ver);
-                future_marker_vrf_proofs.push(
-                    self.vrf
-                        .get_label_proof::<H>(uname, false, ver)
-                        .await?
-                        .to_bytes()
-                        .to_vec(),
-                );
-            }
-
-            Ok(HistoryProof {
-                update_proofs,
-                next_few_vrf_proofs,
-                non_existence_of_next_few,
-                future_marker_vrf_proofs,
-                non_existence_of_future_markers,
-            })
-        } else {
-            match std::str::from_utf8(&username) {
-                Ok(name) => Err(AkdError::Storage(StorageError::NotFound(format!(
-                    "User {} at epoch {}",
-                    name, current_epoch
-                )))),
-                _ => Err(AkdError::Storage(StorageError::NotFound(format!(
-                    "User {:?} at epoch {}",
-                    username, current_epoch
-                )))),
-            }
-        }
-    }
-
-    /// Takes in the current state of the server and a label along with
-    /// a "top" number of key updates to generate a proof for.
-    ///
-    /// If the label is present in the current state,
-    /// this function returns all the values & proof of validity
-    /// up to `top_n_updates` results.
-    pub async fn limited_key_history(
+    pub async fn key_history(
         &self,
-        top_n_updates: usize,
         uname: &AkdLabel,
+        params: HistoryParams,
     ) -> Result<HistoryProof<H>, AkdError> {
         // The guard will be dropped at the end of the proof generation
         let _guard = self.cache_lock.read().await;
 
         let current_azks = self.retrieve_current_azks().await?;
         let current_epoch = current_azks.get_latest_epoch();
         let mut user_data = self.storage.get_user_data(uname).await?.states;
-        // reverse sort from highest epoch to lowest
-        user_data.sort_by(|a, b| b.epoch.partial_cmp(&a.epoch).unwrap());
 
-        let limited_history = user_data
-            .into_iter()
-            .take(top_n_updates)
-            .collect::<Vec<_>>();
+        // reverse sort from highest epoch to lowest
+        user_data.sort_by(|a, b| b.epoch.cmp(&a.epoch));
+
+        // apply filters specified by HistoryParams struct
+        user_data = match params {
+            HistoryParams::Complete => user_data,
+            HistoryParams::MostRecent(n) => user_data.into_iter().take(n).collect::<Vec<_>>(),
+            HistoryParams::SinceEpoch(epoch) => user_data
+                .into_iter()
+                .filter(|val| val.epoch >= epoch)
```

Comment:
Do we want to sort again after the filtering? I am assuming ordering will be maintained, but we should at least check if another sort is necessary.

Reply:
Thanks, I went ahead and added another sorting after the filtering, just in case.